chips/litex_vexriscv/src/interrupt_controller.rs

```diff
@@ -0,0 +1,156 @@
+//! VexRiscv-specific interrupt controller implementation
+
+use core::cell::Cell;
+use core::mem::size_of;
+
+/// Global static variable for the InterruptController, as it must be
+/// accessible to the raw interrupt handler functions
+static mut INTERRUPT_CONTROLLER: VexRiscvInterruptController = VexRiscvInterruptController::new();
+
+/// Rust wrapper around the raw CSR-based VexRiscv interrupt
+/// controller
+///
+/// The wrapper supports saving all currently pending interrupts to an
+/// internal state, which can then be used for interrupt processing.
+pub struct VexRiscvInterruptController {
+    saved_interrupts: Cell<usize>,
+}
+
+impl VexRiscvInterruptController {
+    /// Construct a new VexRiscvInterruptController instance
+    pub const fn new() -> Self {
+        VexRiscvInterruptController {
+            saved_interrupts: Cell::new(0),
+        }
+    }
+
+    /// Reference to the global VexRiscvInterruptController instance
+    ///
+    /// This is safe as long as Tock is guaranteed to only have a
+    /// single thread of execution and the methods don't execute
+    /// external functions (closures, etc.)
+    pub fn global_instance() -> &'static Self {
+        unsafe { &INTERRUPT_CONTROLLER }
+    }
+
+    /// Save the currently pending interrupts in hardware to the
+    /// internal state
+    ///
+    /// This should be accessed in an atomic context to ensure a
+    /// consistent view on the pending interrupts is saved.
+    pub unsafe fn save_pending(&self) -> bool {
+        let all_pending = vexriscv_irq_raw::irq_pending();
+        self.saved_interrupts.set(all_pending);
+
+        // return true if some interrupts were saved
+        all_pending != 0
+    }
+
+    /// Return the next pending interrupts in the saved state
+    ///
+    /// If no interrupt is pending in the saved state, this function
+    /// returns `None`.
+    ///
+    /// The ordering is determined by the interrupt number, lower
+    /// having a higher priority.
+    pub fn next_saved(&self) -> Option<usize> {
+        let saved_interrupts = self.saved_interrupts.get();
+        let interrupt_bits = size_of::<usize>() * 8;
+
+        // This is essentially an inefficient version of C's find first
+        // set (ffs()) function, giving the index of the least significant
+        // bit that is set
+        (0..interrupt_bits).find(|test| saved_interrupts & (1 << test) != 0)
```

Comment:
Wouldn't `leading_zeros()` work here: https://doc.rust-lang.org/std/primitive.usize.html#method.leading_zeros

Reply:
Close! `trailing_zeros` appears to do the job. Unfortunately (and understandably) `usize::trailing_zeros(0) == usize::BITS == core::mem::size_of<usize>() * 8` and because `::BITS` is unstable, we're going to have to use the latter method to check whether there are any interrupts pending. Still worth the potential performance improvements by using LLVM intrinsics.